jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,461 @@
+ * Copyright (c) 2021, 2021 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+package jdk.internal.foreign.abi;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.invoke.WrongMethodTypeException;
+import java.lang.ref.SoftReference;
+import java.lang.invoke.MethodHandles.Lookup;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Arrays;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+
+import jdk.incubator.foreign.Addressable;
+import jdk.incubator.foreign.FunctionDescriptor;
+import jdk.incubator.foreign.MemoryLayout;
+import jdk.incubator.foreign.ValueLayout;
+import jdk.incubator.foreign.MemoryAddress;
+import jdk.incubator.foreign.Addressable;
+import jdk.incubator.foreign.MemorySegment;
+import jdk.incubator.foreign.LibraryLookup;
+import static jdk.incubator.foreign.LibraryLookup.Symbol;
+import jdk.incubator.foreign.CLinker.TypeKind;
+import static jdk.incubator.foreign.CLinker.TypeKind.*;
+
+/**
+ * The counterpart in OpenJDK is replaced with this class that wrap up a method handle
+ * enabling the native code to the ffi_call via the libffi interface at runtime.
+ */
+public class ProgrammableInvoker {
+	
+	private final MethodType funcMethodType;
+	private final FunctionDescriptor funcDescriptor;
+	private final Addressable functionAddr;
+	private long cifNativeThunkAddr;
+	private long argTypesAddr;
+	private List<MemoryLayout> argLayouts;
+	private MemoryLayout[] argLayoutArray;
+	
+	final static Lookup lookup = MethodHandles.lookup();
+	
+	/* The prep_cif and the corresponding argument layouts are cached & shared in multiple downcalls/threads */
+	private static final ConcurrentHashMap<String, Long> cachedCifNativeThunkAddr = new ConcurrentHashMap<String, Long>();
```

Comment:
do you still need these to be concurrentHashMAps if you are using the `privateClassLock`?

Reply:
Updated with `HashMap` and verified with the test cases.